### ISSUE TITLE
[Firestore] Question:  Lifecycle recommendations for FirestoreClient / FirestoreDb

### ISSUE THREAD

**Issue by veatcha** [GoogleCloudPlatform/google-cloud-dotnet#2405]
I'm spinning up my own FirestoreDb/FirestoreClient/Channel/Credential chain as part of initializing our connection to Firestore, as recommended, where our credential management doesn't work with the built-in methods.

I'm wondering if there are recommendations for the lifecycle of the FirestoreClient and FirestoreDb - are they expensive to create?  Can I initialize a singleton for the life of my app, or is it better to spin one up per request?

I ask because we are seeing a small number of requests coming back with this message:

Status(StatusCode=Unknown, Detail="Stream removed")

Which made me wonder if an underlying connection had disappeared or closed...

Thanks,
Andrew

**Comment by jskeet** [comment]
I'd be interested to know more about the "stream removed" issue as a separate matter (perhaps you could file it as a new issue with more details?) but basically: if you only have a single `ChannelCredential` to use, then the recommended approach is indeed to use a singleton for the life of the app. Note that we did have issues with earlier versions of Grpc.Core - if you're not using 1.13.1 it would be worth doing that. I don't *think* it would explain this particular issue, but I could be wrong.

The `FirestoreDb` itself is cheap to create, but you certainly don't want to create a new `Channel` on each request. If you're able to use either a single instance of a limited set (e.g. for a small, known number of credentials) then make those long-lasting and it should be fine.

**Comment by veatcha** [comment]
Thanks @jskeet - We had both as a singleton, so I will keep the Channel/Client there, but make a new Db per usage.

We're on 1.13.1, but I can also bring that up to 1.14, and the Firestore package to beta6.

I will see about collecting some more data on the "steam removed"

**Comment by jskeet** [comment]
Ooh, Grpc.Core 1.14 is *very* new. (I checked yesterday when I rereleased everything for GAX 2.5.0) I've no idea whether that'll help, but it's worth trying.

Updating the Firestore package shouldn't make any change if you're on a recent version - the update yesterday was really just to get all the dependencies up to date. There aren't any recent code changes.

**Comment by veatcha** [comment]
Probably not the best place to ask this, but since you're here - is Retry baked into any of the calls in the Firestore package?  I didn't any mention in the docs, but I did see it in a stack trace...

**Comment by jskeet** [comment]
I'd have to check to be honest - I believe we retry transactions, and we definitely try to "self-heal" stream issues in Watch/Listen API.

**Comment by jtattermusch** [comment]
If you are still able to reproduce the "stream removed" problem, please consider posting additional info in grpc's tracking bug as requested in https://github.com/grpc/grpc/issues/15623#issuecomment-432592307.

**Comment by pblachut** [comment]
Hi, I'm using `Google.Cloud.Firestore` package in version `1.0.0-beta22`. Does anything have changed regarding lifecycle recommendations for FirestoreDb / FirestoreClient especially for backend services?

I use Firestore in backend service to process telemetry data (intensive reads and writes are done there). Service is deployed in GCP GKE and I observe high latency there. I followed all [best practices](https://cloud.google.com/firestore/docs/best-practices) but still 99th percentile is over 200ms (for reads and writes). Such results were present when I had single FirestoreDb instance for each pod.

I achieved lower latency by using basic pool mechanism. Created specified number of FirestoreDbs and then fetch them in random way. All these FirebaseDbs are created using the same json credentials:

```
var client = new FirestoreClientBuilder { JsonCredentials = credentialsJson }.Build();
return FirestoreDb.Create(projectName, client);

Are there any other recommendations for backend services to achieve lower latency?
